### ISSUE TITLE
Looking for documentation on usage for kopf.Patch (with admissioncontroller)

### ISSUE THREAD

**Issue by cjbaar** [nolar/kopf#801]
## Question

I am attempting to build an admission controller, using `on.mutate`. I am having trouble finding detailed documentation or examples of how to use the patch for something other than a top-level key. I have tried the following approaches.
```
def mutate1(name, patch: kopf.Patch, logger, **_):
    """test"""
    logger.info(f"Patching {name}")
    patch.spec['template']['metadata']['annotations']['test-value'] = '123'
```
```
def mutate1(name, patch: kopf.Patch, logger, **_):
    """test"""
    logger.info(f"Patching {name}")
    patch.spec['template']['metadata']['annotations'] = {'test-value': '123'}
```

Neither of these results in the annotation being added to my resource spec. What is the proper way to set a value through nested keys? Also, not sure why both of these approaches appear to fail, but the handler never logs anything other successes.
```
[INFO    ] 192.168.65.3 [15/Jul/2021:17:42:46 +0000] "POST /mutate1?timeout=10s HTTP/1.1" 200 311 "-" "kube-apiserver-admission"
```

## Checklist

- [x ] I have read the [documentation](https://kopf.readthedocs.io/en/latest/) and searched there for the problem
- [x ] I have searched in the [GitHub Issues](https://github.com/nolar/kopf/issues?utf8=%E2%9C%93&q=) for similar questions

## Keywords

patch, webhook

**Comment by cjbaar** [comment]
OK, it looks like I had the handler mis-defined, and it was being filtered out. Got it running, and now I do see actual errors in the logs (webhook still returns 200 though, for some reason). The error is `KeyError: 'spec'`; I assume this comes from the use of `patch.spec`, which is what I see in the docs. So the question is still the same: is there more info on the use of the `patch` object? Thanks.

**Comment by cjbaar** [comment]
FYI, I also tried this:
```
    patch = kopf.Patch({'spec': {'template': {'metadata': {'annotations': {'test-value': '123'} } } } })
```

This doesn't throw an errors in the logs, but also has no effect.

**Comment by nolar** [comment]
The patch is a dict-like object. So, Python's tricks with `patch.setdefault('spec', {}).setdefault('template', {}).setdefault('metadata', {}).setdefault('annotations', {})['xyz'] = 'abc'` should work. All other Python tricks for dicts would work too.

You should use the `patch` object provided by the framework via kwargs. All newly created patch objects have no effect, as the framework is unaware of their existence.

**Comment by cjbaar** [comment]
Thanks. That creates a correct-looking dict, but then the resource fails to create with this message:
`Internal error occurred: replace operation does not apply: doc is missing path: /spec/template/metadata/annotations/xyz: missing value`

Is there an option for the admission controller to use strategic merge patch?

**Comment by cjbaar** [comment]
I have tried checking multiple options down the path. In each case, the `patch` object shows as a correctly-built dict, but it is the same everytime, so the patch method doesn't seem to understand what is being added, and I get the same path error as above.
```
    if not spec.get('template', {}).get('metadata'):
        patch.setdefault('spec', {}).setdefault('template', {})['metadata'] = {'annotations': {'xyz': 'abc'}}
    elif not spec.get('template', {}).get('metadata').get('annotations'):
        patch.setdefault('spec', {}).setdefault('template', {})\
            .setdefault('metadata', {})['annotations'] = {'xyz2': 'abc2'}
    else:
        patch.setdefault('spec', {}).setdefault('template', {}).setdefault('metadata', {})\
            .setdefault('annotations', {})['xyz3'] = 'abc3'
```
```
Internal error occurred: replace operation does not apply: doc is missing path: /spec/template/metadata/annotations/xyz2: missing value
```

How can the patch be used to add a complex key that did not exist before (in this case, `annotations`)?

**Comment by cjbaar** [comment]
After testing on this multiple resource types, I now believe this to be either a bug, or something that was never implemented. I am only able to patch existing items and one layer down. When trying to patch something where more than one key in the chain does not yet exists, it fails. The patch methodology is not able to determine the difference between an `add` and a `replace` operation.

For example, with the following resource spec:
```
spec:
  template:
    metadata:
      labels:
        app: rollouts-demo
```

I can successfully patch `labels` with:
```
    patch.setdefault('spec', {}).setdefault('template', {}).setdefault('metadata', {})\
        .setdefault('labels', {})['xyz'] = 'abc'
```
Which becomes:
```
spec:
  template:
    metadata:
      labels:
        app: rollouts-demo
        xyz: abc
```

But patching `annotations`, which was not present in the initial spec, fails with:
```
    patch.setdefault('spec', {}).setdefault('template', {}).setdefault('metadata', {})\
        .setdefault('annotations', {})['xyz'] = 'abc'
```
Depending on the method used to build the patch data, I get either `KeyError: spec` from the kopf logs, or I get the above error (`doc is missing path`) from the k8s API.

**Comment by cjbaar** [comment]
Well I guess this would explain the issue (https://github.com/nolar/kopf/blob/main/kopf/_cogs/structs/patches.py#L94)
```
# TODO: need to distinguish 'add' vs 'replace' -- need to know the original value.
```

What is the quickest way to get this working? I have been experimenting with just allowing the user to set the patch ops directly, instead of trying to have kopf infer them. Something like this:
```
    def __init__(self, __src: Optional[MutableMapping[str, Any]] = None) -> None:
...
        self._json: List[JSONPatchItem] = []
...
    @property
    def json(self) -> JSONPatch:
        return self._json

    def as_json_patch(self) -> JSONPatch:
        if self._json:
            return self._json
        return [] if not self else self._as_json_patch(self, keys=[''])
```

**Comment by cjbaar** [comment]
@nolar  Not sure how elegant my solution is, but attempted to implemented the `add` operation with the above PR

**Comment by nolar** [comment]
@cjbaar Thank you for fixing this issue!

The fix is merged & pre-released as [`kopf==1.33rc2`](https://github.com/nolar/kopf/releases/tag/1.33rc2) — so that you can already try it. I'll make it into a release presumably on Sunday or Monday 25-26.07.2021.

If there are no other unresolved problems left in this issue, please close it. If there are some, please highlight what is left in a comment.

**Comment by cjbaar** [comment]
@nolar  Thanks! Implementing it now.
